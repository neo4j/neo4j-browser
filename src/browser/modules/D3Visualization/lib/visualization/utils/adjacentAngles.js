/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export default class AdjacentAngles {
  findRuns(AngleList, minSeparation) {
    let p = 0
    let start = 0
    let end = 0
    const runs = []
    const minStart = function() {
      if (runs.length === 0) {
        return 0
      } else {
        return runs[0].start
      }
    }

    const scanForDensePair = function() {
      start = p
      end = AngleList.wrapIndex(p + 1)
      if (end === minStart()) {
        return 'done'
      } else {
        p = end
        if (tooDense(start, end)) {
          return extendEnd
        } else {
          return scanForDensePair
        }
      }
    }

    const extendEnd = function() {
      if (p === minStart()) {
        return 'done'
      } else if (tooDense(start, AngleList.wrapIndex(p + 1))) {
        end = AngleList.wrapIndex(p + 1)
        p = end
        return extendEnd
      } else {
        p = start
        return extendStart
      }
    }

    const extendStart = function() {
      const candidateStart = AngleList.wrapIndex(p - 1)
      if (tooDense(candidateStart, end) && candidateStart !== end) {
        start = candidateStart
        p = start
        return extendStart
      } else {
        runs.push({
          start,
          end
        })
        p = end
        return scanForDensePair
      }
    }

    const tooDense = function(start, end) {
      const run = {
        start,
        end
      }
      return AngleList.angle(run) < AngleList.length(run) * minSeparation
    }

    let stepCount = 0
    let step = scanForDensePair
    while (step !== 'done') {
      if (stepCount++ > AngleList.totalLength() * 10) {
        console.log(
          'Warning: failed to layout arrows',
          (() => {
            const result = []
            for (const key of Object.keys(AngleList.list || {})) {
              const value = AngleList.list[key]
              result.push(`${key}: ${value.angle}`)
            }
            return result
          })().join('\n'),
          minSeparation
        )
        break
      }
      step = step()
    }

    return runs
  }
}
