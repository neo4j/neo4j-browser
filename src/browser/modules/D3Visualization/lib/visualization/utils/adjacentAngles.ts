/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import AngleList from './angleList'

export type Run = { start: number; end: number }

export default class AdjacentAngles {
  findRuns(angleList: AngleList, minSeparation: number): Run[] {
    let p = 0
    let start = 0
    let end = 0
    const runs: Run[] = []
    const minStart = function () {
      if (runs.length === 0) {
        return 0
      } else {
        return runs[0].start
      }
    }

    const scanForDensePair = function () {
      start = p
      end = angleList.wrapIndex(p + 1)
      if (end === minStart()) {
        return 'done'
      } else {
        p = end
        if (tooDense(start, end)) {
          return extendEnd
        } else {
          return scanForDensePair
        }
      }
    }

    const extendEnd = function () {
      if (p === minStart()) {
        return 'done'
      } else if (tooDense(start, angleList.wrapIndex(p + 1))) {
        end = angleList.wrapIndex(p + 1)
        p = end
        return extendEnd
      } else {
        p = start
        return extendStart
      }
    }

    const extendStart = function () {
      const candidateStart = angleList.wrapIndex(p - 1)
      if (tooDense(candidateStart, end) && candidateStart !== end) {
        start = candidateStart
        p = start
        return extendStart
      } else {
        runs.push({
          start,
          end
        })
        p = end
        return scanForDensePair
      }
    }

    const tooDense = function (start: number, end: number) {
      const run = {
        start,
        end
      }
      return angleList.angle(run) < angleList.length(run) * minSeparation
    }

    // Recursive function that's hard to properly type
    let step: 'done' | (() => any) = scanForDensePair

    let stepCount = 0
    while (step !== 'done') {
      if (stepCount++ > angleList.totalLength() * 10) {
        console.log(
          'Warning: failed to layout arrows',
          Object.entries(angleList)
            .map(([key, value]) => `${key}: ${value.angle}`)
            .join('\n'),
          minSeparation
        )
        break
      }
      step = step()
    }

    return runs
  }
}
