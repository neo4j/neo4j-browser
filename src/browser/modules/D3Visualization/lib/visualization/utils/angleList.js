/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export default class AngleList {
  constructor(list) {
    this.list = list
  }

  getAngle(index) {
    return this.list[index].angle
  }

  fixed(index) {
    return this.list[index].fixed
  }

  totalLength() {
    return this.list.length
  }

  length(run) {
    if (run.start < run.end) {
      return run.end - run.start
    } else {
      return run.end + this.list.length - run.start
    }
  }

  angle(run) {
    if (run.start < run.end) {
      return this.list[run.end].angle - this.list[run.start].angle
    } else {
      return 360 - (this.list[run.start].angle - this.list[run.end].angle)
    }
  }

  wrapIndex(index) {
    if (index === -1) {
      return this.list.length - 1
    } else if (index >= this.list.length) {
      return index - this.list.length
    } else {
      return index
    }
  }
}
