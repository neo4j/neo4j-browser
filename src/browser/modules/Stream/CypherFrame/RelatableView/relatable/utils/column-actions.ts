/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 * This file is part of Neo4j.
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import { some } from 'lodash'

import { RelatableAction } from '../relatable.types'
import arrayHasItems from './array-has-items'

export function columnHasActions(
  column: any,
  availableActions: RelatableAction[]
): boolean {
  return (
    arrayHasItems(availableActions) &&
    some(availableActions, action => columnHasAction(column, action))
  )
}

export function columnHasAction(column: any, action: RelatableAction): boolean {
  const [, predicate] = action

  return predicate(column)
}
