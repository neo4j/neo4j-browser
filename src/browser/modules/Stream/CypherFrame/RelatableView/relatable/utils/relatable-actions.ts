/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 * This file is part of Neo4j.
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import { filter, find, map, some } from 'lodash'

import { RelatableAction, TableAddOnReturn } from '../relatable.types'

export function getRelatableGlobalActions(
  addOns: TableAddOnReturn[]
): RelatableAction[] {
  return map(
    filter(addOns, ([name]) => Boolean(name)),
    ([name, , predicate]): RelatableAction => [name!, predicate]
  )
}

export function getRelatableTableActions(
  addOns: TableAddOnReturn[]
): RelatableAction[] {
  return map(
    filter(addOns, ([, name]) => Boolean(name)),
    ([, name, predicate]): RelatableAction => [name!, predicate]
  )
}

export function getRelatableAction(
  actions: RelatableAction[],
  name: string
): RelatableAction | undefined {
  return find(actions, ([actionName]) => actionName === name)
}

export function isActionAvailable(
  availableActions: RelatableAction[],
  name: string
): boolean {
  return some(availableActions, ([actionName]) => actionName === name)
}
