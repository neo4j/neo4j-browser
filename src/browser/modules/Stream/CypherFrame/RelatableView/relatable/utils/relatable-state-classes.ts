/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 * This file is part of Neo4j.
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import { entries, kebabCase, keys, map, reduce } from 'lodash'

import { ROW_STATE_CLASSES, TOOLBAR_STATE_CLASSES } from '../constants'

export function getRowClasses(row: any) {
  return `relatable__table-row relatable__table-body-row ${getRowStateClasses(
    row
  )}`
}

export function createRowStateClasses() {
  return map(entries(ROW_STATE_CLASSES), ([state, bgColor]) =>
    bgColor
      ? `.${getRowStateClass(state)} {
        background-color: ${bgColor};
      }`
      : ''
  )
}

export function createToolbarStateClasses() {
  return map(
    entries(TOOLBAR_STATE_CLASSES),
    ([state, bgColor]) => `
    /* need to win over semantic specificity */
    .menu .label.${getToolbarStateClass(state)} {
      background-color: ${bgColor};
    }
  `
  )
}

export function getToolbarStateClass(state: string) {
  return `relatable__toolbar-label--${kebabCase(state)}`
}

function getRowStateClasses(row: any) {
  return reduce(
    keys(ROW_STATE_CLASSES),
    (agg, state) => (row[state] ? `${agg} ${getRowStateClass(state)}` : agg),
    ''
  )
}

function getRowStateClass(state: string) {
  return `relatable__table-row-state--${kebabCase(state)}`
}
