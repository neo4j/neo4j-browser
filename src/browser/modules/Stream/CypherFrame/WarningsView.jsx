/*
 * Copyright (c) 2002-2018 "Neo4j, Inc"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { Component } from 'preact'
import { deepEquals } from 'services/utils'
import {
  StyledCypherMessage,
  StyledCypherWarningMessage,
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledBr,
  StyledHelpFrame
} from '../styled'

const getWarningComponent = severity => {
  if (severity === 'ERROR') {
    return <StyledCypherErrorMessage>{severity}</StyledCypherErrorMessage>
  } else if (severity === 'WARNING') {
    return <StyledCypherWarningMessage>{severity}</StyledCypherWarningMessage>
  } else {
    return <StyledCypherMessage>{severity}</StyledCypherMessage>
  }
}

export class WarningsView extends Component {
  shouldComponentUpdate (props, state) {
    if (!this.props.result) return true
    return !deepEquals(props.result.summary, this.props.result.summary)
  }
  render () {
    if (this.props.result === undefined) return null
    const { summary = {} } = this.props.result
    const { notifications = [], statement = {} } = summary
    const { text: cypher = '' } = statement
    if (!notifications || !cypher) {
      return null
    }

    let cypherLines = cypher.split('\n')
    cypherLines[0] = cypherLines[0].replace(/^EXPLAIN /, '')

    let notificationsList = notifications.map(notification => {
      return (
        <StyledHelpContent>
          <StyledHelpDescription>
            {getWarningComponent(notification.severity)}
            <StyledH4>{notification.title}</StyledH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledHelpDescription>
              {notification.description}
            </StyledHelpDescription>
            <StyledDiv>
              <StyledPreformattedArea>
                {cypherLines[notification.position.line - 1]}
                <StyledBr />
                {Array(notification.position.column).join(' ')}^
              </StyledPreformattedArea>
            </StyledDiv>
          </StyledDiv>
        </StyledHelpContent>
      )
    })
    return <StyledHelpFrame>{notificationsList}</StyledHelpFrame>
  }
}

export class WarningsStatusbar extends Component {
  shouldComponentUpdate (props, state) {
    return false
  }
  render () {
    return null
  }
}
