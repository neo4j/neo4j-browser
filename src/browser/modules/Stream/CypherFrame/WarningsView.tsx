/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react'
import { deepEquals } from 'services/utils'
import {
  StyledCypherMessage,
  StyledCypherWarningMessage,
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledBr,
  StyledHelpFrame
} from '../styled'

const getWarningComponent = (severity: any) => {
  if (severity === 'ERROR') {
    return <StyledCypherErrorMessage>{severity}</StyledCypherErrorMessage>
  } else if (severity === 'WARNING') {
    return <StyledCypherWarningMessage>{severity}</StyledCypherWarningMessage>
  } else {
    return <StyledCypherMessage>{severity}</StyledCypherMessage>
  }
}

export class WarningsView extends Component<any> {
  shouldComponentUpdate(props: any) {
    if (!this.props.result) return true
    return !deepEquals(props.result.summary, this.props.result.summary)
  }

  render() {
    if (this.props.result === undefined) return null
    const { summary = {} } = this.props.result
    const { notifications = [], query = {} } = summary
    const { text: cypher = '' } = query
    if (!notifications || !cypher) {
      return null
    }
    const cypherLines = cypher.split('\n')
    const notificationsList = notifications.map((notification: any) => {
      // Detect generic warning without position information
      const position = Object.keys(notification.position).length
        ? notification.position
        : { line: 1, offset: 0 }
      return (
        <StyledHelpContent
          key={notification.title + position.line + position.offset}
        >
          <StyledHelpDescription>
            {getWarningComponent(notification.severity)}
            <StyledH4>{notification.title}</StyledH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledHelpDescription>
              {notification.description}
            </StyledHelpDescription>
            <StyledDiv>
              <StyledPreformattedArea>
                {cypherLines[position.line - 1]}
                <StyledBr />
                {Array(position.offset + 1).join(' ')}^
              </StyledPreformattedArea>
            </StyledDiv>
          </StyledDiv>
        </StyledHelpContent>
      )
    })
    return <StyledHelpFrame>{notificationsList}</StyledHelpFrame>
  }
}

export class WarningsStatusbar extends Component<any> {
  shouldComponentUpdate() {
    return false
  }

  render() {
    return null
  }
}
