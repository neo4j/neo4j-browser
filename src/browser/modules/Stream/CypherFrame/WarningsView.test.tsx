/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { render } from '@testing-library/react'
import React from 'react'

import { WarningsView, WarningsViewProps } from './WarningsView'
import { Provider } from 'react-redux'

import { initialState as initialMetaState } from 'shared/modules/dbMeta/dbMetaDuck'
import { initialState as initialSettingsState } from 'shared/modules/settings/settingsDuck'
import { createBus } from 'suber'
import { DeepPartial } from 'shared/utils/deepPartial'
import { notificationFilterMinimumSeverityLevel } from 'neo4j-driver-core'

const withProvider = (store: any, children: any) => {
  return <Provider store={store}>{children}</Provider>
}

const mount = (props: DeepPartial<WarningsViewProps>, state?: any) => {
  const defaultProps: WarningsViewProps = {
    result: null,
    bus: createBus()
  }

  const combinedProps = {
    ...defaultProps,
    ...props
  }

  const initialState = {
    app: {},
    meta: initialMetaState,
    settings: initialSettingsState
  }

  const combinedState = { ...initialState, ...state }

  const store = {
    subscribe: () => {},
    dispatch: () => {},
    getState: () => ({
      ...combinedState
    })
  }

  return render(withProvider(store, <WarningsView {...combinedProps} />))
}

describe('WarningsView', () => {
  test('displays nothing if no notifications', () => {
    // Given
    const props = {
      result: null
    }

    // When
    const { container } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display a warning', () => {
    // Given
    const props = {
      result: {
        summary: {
          notifications: [
            {
              severity: 'WARNING',
              title: 'My xx1 warning',
              description: 'This is xx2 warning',
              position: {
                offset: 7,
                line: 1
              },
              code: 'xx3.Warning'
            }
          ],
          query: {
            text: 'EXPLAIN MATCH xx3'
          }
        }
      }
    }

    // When
    const { container } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display a warning for GQL status codes', () => {
    // Given
    const props = {
      result: {
        summary: {
          server: {
            protocolVersion: 5.6
          },
          gqlStatusObjects: [
            {
              severity: notificationFilterMinimumSeverityLevel.WARNING,
              gqlStatus: '03N90',
              statusDescription:
                "info: cartesian product. The disconnected pattern 'p = ()--(), q = ()--()' builds a cartesian product. A cartesian product may produce a large amount of data and slow down query processing.",
              position: {
                offset: 7,
                line: 1
              }
            }
          ],
          query: {
            text: 'MATCH p=()--(), q=()--() RETURN p, q'
          }
        }
      }
    }

    const state = {
      meta: {
        server: {
          version: '5.27.0'
        }
      }
    }

    // When
    const { container } = mount(props, state)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display multiple warnings', () => {
    // Given
    const props = {
      result: {
        summary: {
          notifications: [
            {
              severity: 'WARNING',
              title: 'My xx1 warning',
              description: 'This is xx2 warning',
              position: {
                offset: 7,
                line: 1
              },
              code: 'xx3.Warning'
            },
            {
              severity: 'WARNING',
              title: 'My yy1 warning',
              description: 'This is yy2 warning',
              position: {
                offset: 3,
                line: 1
              },
              code: 'yy3.Warning'
            }
          ],
          query: {
            text: 'EXPLAIN MATCH zz3'
          }
        }
      }
    }

    // When
    const { container } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display multiple warnings for GQL status codes', () => {
    // Given
    const props = {
      result: {
        summary: {
          server: {
            protocolVersion: 5.6
          },
          gqlStatusObjects: [
            {
              severity: notificationFilterMinimumSeverityLevel.WARNING,
              gqlStatus: '03N90',
              statusDescription:
                "info: cartesian product. The disconnected pattern 'p = ()--(), q = ()--()' builds a cartesian product. A cartesian product may produce a large amount of data and slow down query processing.",
              position: {
                offset: 7,
                line: 1
              }
            },
            {
              severity: notificationFilterMinimumSeverityLevel.WARNING,
              gqlStatus: '01N50',
              statusDescription:
                'warn: label does not exist. The label `A` does not exist. Verify that the spelling is correct.',
              position: {
                offset: 3,
                line: 1
              }
            }
          ],
          query: {
            text: 'MATCH p=()--(), q=()--() RETURN p, q'
          }
        }
      }
    }

    const state = {
      meta: {
        server: {
          version: '5.27.0'
        }
      }
    }

    // When
    const { container } = mount(props, state)

    // Then
    expect(container).toMatchSnapshot()
  })
})
