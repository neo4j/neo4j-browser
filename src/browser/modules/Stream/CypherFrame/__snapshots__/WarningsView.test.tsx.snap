// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningsViews WarningsView displays nothing if no notifications 1`] = `<div />`;

exports[`WarningsViews WarningsView does displays a warning 1`] = `
<div>
  <div
    class="sc-hGPBjI jLpNLK"
  >
    <div
      class="sc-dlVxhl fwhZkz"
    >
      <div
        class="sc-kfPuZi lgMZCo"
      >
        <div
          class="sc-cxpSdN cnlYxc"
        >
          WARNING xx0
        </div>
        <h4
          class="sc-kLwhqv"
        >
          My xx1 warning
        </h4>
      </div>
      <div
        class="sc-fKVqWL"
      >
        <div
          class="sc-kfPuZi lgMZCo"
        >
          This is xx2 warning
        </div>
        <div
          class="sc-fKVqWL"
        >
          <pre
            class="sc-hiCibw jvxLJD"
          >
            EXPLAIN MATCH xx3
            <br
              class="sc-jJoQJp"
            />
                   
            ^
          </pre>
        </div>
      </div>
      <div
        class="sc-fKVqWL"
        style="margin-top: 10px;"
      >
        Status code: 
        <code
          class="sc-ZOtfp koHcGf"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`WarningsViews WarningsView does displays multiple warnings 1`] = `
<div>
  <div
    class="sc-hGPBjI jLpNLK"
  >
    <div
      class="sc-dlVxhl fwhZkz"
    >
      <div
        class="sc-kfPuZi lgMZCo"
      >
        <div
          class="sc-cxpSdN cnlYxc"
        >
          WARNING xx0
        </div>
        <h4
          class="sc-kLwhqv"
        >
          My xx1 warning
        </h4>
      </div>
      <div
        class="sc-fKVqWL"
      >
        <div
          class="sc-kfPuZi lgMZCo"
        >
          This is xx2 warning
        </div>
        <div
          class="sc-fKVqWL"
        >
          <pre
            class="sc-hiCibw jvxLJD"
          >
            EXPLAIN MATCH zz3
            <br
              class="sc-jJoQJp"
            />
                   
            ^
          </pre>
        </div>
      </div>
      <div
        class="sc-fKVqWL"
        style="margin-top: 10px;"
      >
        Status code: 
        <code
          class="sc-ZOtfp koHcGf"
        />
      </div>
    </div>
    <div
      class="sc-dlVxhl fwhZkz"
    >
      <div
        class="sc-kfPuZi lgMZCo"
      >
        <div
          class="sc-cxpSdN cnlYxc"
        >
          WARNING yy0
        </div>
        <h4
          class="sc-kLwhqv"
        >
          My yy1 warning
        </h4>
      </div>
      <div
        class="sc-fKVqWL"
      >
        <div
          class="sc-kfPuZi lgMZCo"
        >
          This is yy2 warning
        </div>
        <div
          class="sc-fKVqWL"
        >
          <pre
            class="sc-hiCibw jvxLJD"
          >
            EXPLAIN MATCH zz3
            <br
              class="sc-jJoQJp"
            />
               
            ^
          </pre>
        </div>
      </div>
      <div
        class="sc-fKVqWL"
        style="margin-top: 10px;"
      >
        Status code: 
        <code
          class="sc-ZOtfp koHcGf"
        />
      </div>
    </div>
  </div>
</div>
`;
